src/signed_stubs.c

```diff
@@ -0,0 +1,143 @@
+#define __USE_MINGW_ANSI_STDIO 1
+#endif
+
+#include <caml/mlvalues.h>
+#include <caml/custom.h>
+#include <caml/alloc.h>
+#include <caml/intext.h>
+#include <caml/fail.h>
+
+#include <inttypes.h>
+#include <stdint.h>
+#include <limits.h>
+#include <stdio.h>
+
+#define OCAML_INTEGERS_INTERNAL 1
+#include "ocaml_integers.h"
+
+CAMLprim value integers_int_size(value unit)
+{
+  return Val_long(8 * sizeof(value) - 1) ;
+}
+
+#define INT_SMALL_DECLS(BITS)                                               \
+  /* of_string : string -> t */                                             \
+  extern value integers_int ## BITS ## _of_string(value a);                 \
+  /* to_string : t -> string */                                             \
+  extern value integers_int ## BITS ## _to_string(value a);                 \
+  /* max : unit -> t */                                                     \
+  extern value integers_int ## BITS ## _max(value a);
+
+INT_SMALL_DECLS(8)
+INT_SMALL_DECLS(16)
+
+#define TYPE(SIZE) int ## SIZE ## _t
+#define UTYPE(SIZE) uint ## SIZE ## _t
+#define BUF_SIZE(TYPE) ((sizeof(TYPE) * CHAR_BIT + 2) / 3 + 2)
+
+#define INT_OF_STRING(BITS, COPY)                                            \
+  value integers_int ## BITS ## _of_string(value a)                          \
+  {                                                                          \
+    TYPE(BITS) i, max_prefix;                                                \
+    const char *pos = String_val(a);                                         \
+    int base = 10, d, sign = 1, signedness = 1;                              \
+                                                                             \
+    /* Detect sign, if given */                                              \
+    if (*pos == '-') {                                                       \
+       sign = -1;                                                            \
+      pos++;                                                                 \
+    } else if (*pos == '+') {                                                \
+      pos++;                                                                 \
+    }                                                                        \
+    if (*pos == '0') {                                                       \
+      switch (pos[1]) {                                                      \
+        case 'x': case 'X':                                                  \
+          base = 16; pos += 2; break;                                        \
+        case 'o': case 'O':                                                  \
+          base = 8; pos += 2; break;                                         \
+        case 'b': case 'B':                                                  \
+          base = 2; pos += 2; break;                                         \
+        case 'u': case 'U': /* Unsigned prefix. No-op for unsigned types */  \
```

Comment:
I left this comment in place because I'm still not 100% sure what the behavior here should be after reading the comments in `Stdlib.int_of_string` -- I think it's possible this case is already taken care of here since we're parsing directly into an integer of correct size rather than an `intnat`?